Display build number in summary and support offsetting of Riff Raff Build number

paulmr opened the pull request:
## What does this change?

This change address what I believe is a common use case, where a project has switched from a previous build tool to GitHub actions, and we want the GitHub Actions Riff-Raff build number to start counting from the last build number in the previous tool.

This also adds a "Riff Raff" section to the job summary which outputs the build number selected (whether or not it has been modified).

<!-- A PR should have enough detail to be understandable far in the future. e.g what is the problem/why is the change needed, how does it solve it and any questions or points of discussion. Prefer copying information from a Trello card over linking to it; the card may not always exist and reviewers may not have access to the board. -->

## How to test

If you provide the optional `buildNumberOffset` argument when calling this action, that value should be added to the build number (e.g. in the default case, without also providing a `buildNumber` parameter, this would get added to the GitHub Run Number).

If this optional parameter is not provided then the build number should remain unchanged.

In either case, after running the action, you should be able to click on the run to see it's summary, and at the bottom you will see a Riff Raff section outputting the build number.

An example of where I have run this on an actual build can be seen here:

https://github.com/guardian/kindle-publication-check/actions/runs/3439230129

## Images

This is an example of the output added to the job summary:

<img width="438" alt="image" src="https://user-images.githubusercontent.com/2242307/201172475-bc0daef4-977c-4491-963f-4e45df158f19.png">

akash1810 commented on src/config.ts:
```diff
@@ -129,16 +129,35 @@ export interface Configuration {
 
+const offsetBuildNumber = (
+	buildNumber: string,
+	offset: string | undefined,
```
Wonder if we can simplify this by defaulting offset to 0?

paulmr replied:
yes good point. I am not a typescript expert (far from it!), can functions have default arguments?

akash1810 commented on src/config.ts:
```diff
@@ -129,16 +129,35 @@ export interface Configuration {
+	if (isNaN(intOffset) || isNaN(intBuildNumber)) {
+		return buildNumber;
+	} else {
+		return `${intBuildNumber + intOffset}`;
```
Gosh! Could we use the `number` type here? Whilst JS _can_ add two number strings, it shouldn't! 😂

paulmr replied:
I am actually not sure I understand your comment here. I think `intBuildNumber` and `intOffset` are both of type `number` because they are initialised by `parseInt()`, which returns a `number`, is that not right?

We need to convert them from a string at some point because my understanding is that all input args to an action are strings, and also it looks like we support non-numeric build numbers any way as the default value is actually `dev` for the build number (which seems fair enough). That's why I made sure to check for `NaN`.

akash1810 replied:
🤦🏽  yep, you're right!

minor: Maybe this'll be more explicit:

```ts
return (intBuildNumber + intOffset).toString()
```

WDYT?

paulmr replied:
Oh yeah that is much better actually!

paulmr commented:
Thanks a lot reviewing Akash!

> This is excellent! Wonder if we should add more detail to the table? e.g. the Riff-Raff project name, a deep-link to Riff-Raff to deploy to CODE?

Yes, I love this idea. I will admit I was being a little bit sheepish about adding too much to the summary in case we weren't happy to make a big deal out of it, but I think this is definitely a good idea to add more info.

paulmr commented:
I have updated this to default the offset to zero, and also added the project name to the build summary:

<img width="548" alt="image" src="https://user-images.githubusercontent.com/2242307/201202916-f0389b9f-f4ad-40cf-8354-2d7c3951113e.png">

nicl reviewed (approved):
This is great :) will save a lot of time for people when migrating projects off Teamcity. And the table summary looks really helpful too.

akash1810 commented on action.yaml:
```diff
@@ -19,6 +19,9 @@ inputs:
   buildNumber:
     description: If set, will be used for the build number. (Typically this is read from process.env.GITHUB_RUN_NUMBER.)
     required: false
+  buildNumberOffset:
```
Should we document this input in the [README](https://github.com/guardian/actions-riff-raff/blob/main/README.md#available-inputs)?

paulmr replied:
Yes absolutely, my oversight.

akash1810 commented:
I've made the v2.0.1 tag for this now, and updated the v2 tag. Any projects using `guardian/actions-riff-raff@v2` should see this wonderful annotation.
